### PR TITLE
fix: storage hashing stage gets stuck in a loop on big contracts

### PR DESCRIPTION
**Pull request by joshieDo**
### Motivation
As it was before, if the number of slots of a contract was bigger than `commit_threshold`, the stage would get stuck. 
1. The walker takes `commit_threshold`.
2. `let next_key = storage.next()?;` would give out the same key, since there's still more to go through
3. A new cursor and new walker would be initiated with the same key, but it would start from the beginning again. Back to 1. and so on.

### Solution
Save the cursor position on the last element, and set it for the next iteration.

Did a small refactor w/ address caching.

### CODE REVIEW

File: crates/stages/src/stages/hashing_storage.rs [hunk 1]
```diff
@@ -62,38 +63,72 @@ impl<DB: Database> Stage<DB> for StorageHashingStage {
...
+                                    // Address caching for the first iteration when current_key is
+                                    // None
+                                    let keccak_address =
+                                        if let Some(keccak_address) = keccak_address {
+                                            keccak_address
+                                        } else {
+                                            keccak256(address)
+                                        };
+
+                                    // TODO cache map keccak256(slot.key) ?
```
Comment (gakonst) [inline crates/stages/src/stages/hashing_storage.rs:90]: agreed

**Review by gakonst — CHANGES_REQUESTED** [review]
logic looks good - let's add a test for the missing edge case where commit_threshold < num storage slots

**Comment by joshieDo** [comment]
modified the existing test to cover it. cherry-picking it to main made it hang on that test as expected

File: crates/stages/src/stages/hashing_storage.rs (hunk 1 above)
Comment (joshieDo) [reply crates/stages/src/stages/hashing_storage.rs:90]: but i feel it only makes sense if the cache hashing algo is so much faster than keccak256. 

i'd leave it for the benchmark PR and try it out there

ref #1301

File: crates/stages/src/stages/hashing_storage.rs [hunk 2]
```diff
@@ -62,38 +63,72 @@ impl<DB: Database> Stage<DB> for StorageHashingStage {
...
-                let next_key = {
+                let mut hashed_batch = BTreeMap::new();
+                let mut remaining = self.commit_threshold as usize;
+                {
                     let mut storage = tx.cursor_dup_read::<tables::PlainStorageState>()?;
+                    while !remaining.is_zero() {
+                        hashed_batch.extend(
+                            storage.walk_dup(current_key, current_subkey)?.take(remaining).map(
+                                |res| {
+                                    let (address, slot) = res.expect("should not fail");
```
Comment (joshieDo) [inline crates/stages/src/stages/hashing_storage.rs:79]: fine or should it be changed to accomodate `?` ?

**Review by gakonst — APPROVED** [review]
